### DISCUSSION TITLE
Aff: Problems with type inference

### DISCUSSION THREAD

**Discussion by dirkil** [louthy/language-ext#1335] [Q&A] [answered]
Hi there,

I am currently collecting experience with programming with side effects and hence with `Eff `and `Aff`. I think I am getting the general idea but nevertheless I am struggling and some specific points and would be more than happy if someone could point me in the right direction. 

**Problem #1:**
I have assembled a little example that shows my point. I have a `Processor` class that receives a `List<Customer>` which it should process and sends the result afterwards. Each action can obviously fail. Since the result for each customer is sent to someone, there is no overall result of the `Start` method except when there was a failure. So basically the structure is like that:
```csharp
public Aff<IEnumerable<Unit>> Start(List<Customer> customers)
{
        return customers.Map(cust =>
                from _1 in LogInformation($"Processing {cust.Name}")
                from result in ProcessCustomer(cust)
                from _2 in SendResult(result)
                select unit)
            .Sequence();
}

Aff<Unit> LogInformation(string msg){}

Aff<ProcessResult> ProcessCustomer(Customer customer){}

Aff<Unit> SendResult(ProcessResult result){}
```
Unfortunately, the compiler gives me an error telling me that for the `Sequence` call the types cannot be inferred. I tried a few things but I cannot get it to work. So, if someone sees the problem I would be more than happy to be enlightened.

**Problem #2:**

The `ProcessCustomer` method has to deal with an `Option<string>` and also here the compiler does not like what I am doing. It says **Arguments mismatch** but I cannot see where it goes wrong.

```csharp
public class Processor
{    
    
    private Aff<ProcessResult> ProcessCustomer(Customer customer)
    {
        return
            from inputFile in LoadInputFile(customer.Id)
            from result in inputFile.Match(     // <- Here it goes wrong
                Succ: file => ProcessFile(file), 
                None: _ => NoFileFound(customer)
                )
            select result;
    }

    private Aff<ProcessResult> ProcessFile(string file) =>
        from _1 in LogInformation($"Process {file}")
        select new ProcessResult();

    private Aff<ProcessResult> NoFileFound(Customer customer) =>
        from _1 in LogInformation($"No file found for {customer.Name}")
        select new ProcessResult();

    private Aff<Option<string>> LoadInputFile(int id)
    {
        return default;
    }
    
    private Aff<Unit> LogInformation(string msg) => 
        Aff<Unit>(() => default);   // Log Info somehow...
}

public class ProcessResult
{
    public string Description { get; set; } = string.Empty;
}

public class Customer
{
    public int Id { get; set; }
    
    
    public string Name { get; set; }
}
```

So, if someone could help me to get this compiled I would really appreciate it.

**Comment by harrhp** [comment, marked as answer]
1. There is no `Sequence` overload for `IEnumerable<Aff<T>>`. Use `SequenceSerial` or `SequenceParallel`
2. First argument's name of `Option.Match` is `Some` not `Succ`

**Reply by dirkil** [reply]
Thank you harrhp, 

your answer was spot on. Thank you for pointing me out to `SequenceSerial` and `SequenceParallel`. 

I can't believe that I used `Succ` instead of `Some`. Kind of embarassing... ;)
